src/NuGet.Clients/NuGet.SolutionRestoreManager/VulnerablePackagesInfoBar.cs

```diff
@@ -0,0 +1,163 @@
+                }
+            }
+            finally
+            {
+                _closeFromHide = false;
+            }
+        }
+
+        protected InfoBarModel GetInfoBarModel()
+        {
+            return new InfoBarModel(
+                Resources.InfoBar_TextMessage,
+                new IVsInfoBarActionItem[] {
+                    new InfoBarHyperlink(Resources.InfoBar_HyperlinkMessage),
+                },
+                KnownMonikers.StatusWarning);
+        }
+
+        internal async Task CreateAndShowInfoBarAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (_closed || _visible)
+            {
+                return;
+            }
+
+            IVsInfoBarHost? infoBarHost;
+            try
+            {
+                IVsUIShell? uiShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>(throwOnFailure: false);
+                if (uiShell == null)
+                {
+                    throw new NullReferenceException(nameof(uiShell));
+                }
+
+                if (ErrorHandler.Failed(uiShell.FindToolWindow((uint)__VSFINDTOOLWIN.FTW_fFindFirst, VSConstants.StandardToolWindows.SolutionExplorer, out var windowFrame)))
+                {
+                    throw new NullReferenceException(nameof(windowFrame));
+                }
+
+                object tempObject;
+                if (ErrorHandler.Failed(windowFrame.GetProperty((int)__VSFPROPID7.VSFPROPID_InfoBarHost, out tempObject)))
+                {
+                    throw new NullReferenceException(nameof(tempObject));
+                }
+
+                infoBarHost = (IVsInfoBarHost)tempObject;
+            }
+            catch (Exception ex)
+            {
+                await TelemetryUtility.PostFaultAsync(ex, nameof(VulnerablePackagesInfoBar));
+                return;
+            }
+
+            try
+            {
+                IVsInfoBarUIFactory? infoBarFactory = await _asyncServiceProvider.GetServiceAsync<SVsInfoBarUIFactory, IVsInfoBarUIFactory>(throwOnFailure: false);
+                if (infoBarFactory == null)
+                {
+                    NullReferenceException exception = new NullReferenceException(nameof(infoBarFactory));
+                    await TelemetryUtility.PostFaultAsync(exception, nameof(VulnerablePackagesInfoBar));
+                    return;
+                }
+
+                InfoBarModel infoBarModel = GetInfoBarModel();
+
+                _infoBarUIElement = infoBarFactory.CreateInfoBar(infoBarModel);
+                _infoBarUIElement.Advise(this, out uint cookie);
+                _eventCookie = cookie;
+
+                infoBarHost.AddInfoBar(_infoBarUIElement);
+
+                _visible = true;
+            }
+            catch
+            {
+                return;
+            }
```

Comment:
What about adding fault telemetry as well?

Reply:
> We should be more explicit about what we expect to go wrong and handle it.

Generally good advice, but I think it's risky when calling VS APIs. If the VS component happens to be written in C++, we'll get COM interop exceptions. If in the future the component gets rewritten in .NET, then we'll start getting different exceptions.

In my opinion, when we're just being defensive about "this probably should never fail, but just in case let's fail gracefully and record fault telemetry", then catching all exception is fine. But I agree, if we don't expect exceptions, then we should record telemetry.